### PR TITLE
Password and username validation utility functions

### LINKED ISSUE #15: Utility functions for user creation, password salting
This will be entirely independent of user models (#12) and the database - the goal is to create a `crypto` package and a few utility functions that can be used to run basic validation on a few things:

- emails
- passwords
  - length
  - character legality

There should also be functions to:

- take a password and salt it
- take a password and validate it against a salted password

I implemented pretty complete example of most of this functionality in Inertia that you can refer to:

https://github.com/ubclaunchpad/inertia/blob/master/daemon/inertiad/crypto/password.go

For salting, I think [bcrypt](https://en.wikipedia.org/wiki/Bcrypt) is probably the most elegant solution.

This will be part of the `gateway` sub-project.

### PR DESCRIPTION
**Pull request by omarkhs**
:tickets: **Ticket(s)**: Closes #15 

---

## :construction_worker: Changes

Added crypto package which deals with password hashing, username and password validation to make sure user registers with a legal password and username, and comparing input password with hashed password.

## :flashlight: Testing Instructions
Just run password_test.go. All tests should pass!

### CODE REVIEW

File: core/crypto/password.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+func IsLegalUserName(username string) bool {
+	for _, c := range username {
+		if (c < 'a' || c > 'z') && (c < 'A' || c > 'Z') && (c < 48 || c > 57) && c != '_' && c != '-' {
+			return false
+		}
+	}
+	return true
+}
+
+// IsLegalPassword returns true if the chosen password does not contain illegal characters
```
[the remark this answers is not in the document]
Comment (bobheadxi) [reply core/crypto/password.go]: I think we can implement this a little further down the line - I'll leave complexity enforcement up to @omarkhs

File: core/crypto/password_test.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	correct := comparePasswords(hashed, unhashed)
+	assert.True(t, correct)
+
+	correct = comparePasswords(hashed, "LaunchPadpinpoint")
+	assert.False(t, correct)
+}
+
+func TestValidateCredentialValues(t *testing.T) {
+	err := ValidateCredentialValues("finasdfsdfe", "okaasdfasdy")
+	assert.Nil(t, err)
```
Comment (bobheadxi) [inline core/crypto/password_test.go]: this should be written as table-driven tests, for example:

```go
type args struct {
		opts RunOpts
	}
	tests := []struct {
		name string
		args args
	}{
		{"no options", args{RunOpts{}}},
		{"with core TLS", args{RunOpts{
			CoreOpts: CoreOpts{
				Host:     "localhost",
				Port:     "9111",
				CertFile: "../../dev/certs/127.0.0.1.crt",
			},
		}}},
		{"with gateway TLS", args{RunOpts{
			GatewayOpts: GatewayOpts{
				CertFile: "../../dev/certs/127.0.0.1.crt",
				KeyFile:  "../../dev/certs/127.0.0.1.key",
			},
		}}},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			api, err := New(l)
			if err != nil {
				t.Error(err)
				return
			}
			go api.Run("localhost", "8080", tt.args.opts)
			time.Sleep(time.Millisecond)
			api.Stop()
		})
	}
```

you can generate these in VS code!

File: core/crypto/password.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+var (
+	errSameUsernamePassword = errors.New("Username and password must be different")
+	errInvalidUsername      = errors.New("Username must be at least 3 characters. Only alphanumeric characters, underscores, and dashes are allowed")
+	errInvalidPassword      = errors.New("Password must be at least 5 characters. Only alphanumeric characters, and symbols are alowed")
+)
+
+// hashAndSalt hashes and salts the given user password 
+func hashAndSalt(password string) (string, error) {
+
+	bytePwd := []byte(password)
```
Comment (srijonsaha) [inline core/crypto/password.go]: spacing gone wrong? 😱

**Review by bobheadxi — APPROVED** [review]
Looks great, awesome work!

**Pull request merged**